### ISSUE TITLE
random crashes at emscripten startup... (difficult to reproduce)

### ISSUE THREAD

**Issue by terop2** [emscripten-core/emscripten#20329]
Uncaught (in promise) RuntimeError: Aborted(No EM_ASM constant found at address 2387855)
    at abort (web_page_highmem.js?1695434054:970:11)
    at runMainThreadEmAsm (web_page_highmem.js?1695434054:4283:45)
    at _emscripten_asm_const_int_sync_on_main_thread (web_page_highmem.js?1695434054:4287:14)
    at Emscripten_ShowCursor (web_page_highmem.wasm?1695434054:0x510373)
    at SDL_SetMouseFocus (web_page_highmem.wasm?1695434054:0x4d037d)
    at Emscripten_CreateWindow (web_page_highmem.wasm?1695434054:0x51197d)
    at SDL_CreateWindow (web_page_highmem.wasm?1695434054:0x513af0)
    at SDLApi::SDL_CreateWindow(char const*, int, int, int, int, unsigned int) (web_page_highmem.wasm?1695434054:0x125536)
    at InitSDL2(int, int, bool, bool, bool, bool) (web_page_highmem.wasm?1695434054:0x11f116)
    at GameApi::MainLoopApi::init_window(int, int, std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char>>, bool) (web_page_highmem.wasm?1695434054:0x10d82a)

**Comment by terop2** [comment]
oh, forgot to mention that I'm currently in 1.3.30, but I think its rare enough crash that it has not been fixed yet in newer builds.

**Comment by sbc100** [comment]
This looks like an error that should effect any/all SDL2 users (or rather anyone who calls  SDL_CreateWindow.

Can you share the full list of link flags you are using? 

What does the asmConsts map looks like in the generated JS file?   Does it looks like it contains the EM_ASM blocks from ShowCursor: https://github.com/libsdl-org/SDL/blob/e1789b320edbd8644851bd793875baf32d3b31f6/src/video/emscripten/SDL_emscriptenmouse.c#L196-L212

**Comment by terop2** [comment]
The link flags: (cd ..;em++ -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=30 -fno-rtti -fno-exceptions -s FILESYSTEM=0 -s TOTAL_STACK=400kb -s DISABLE_EXCEPTION_CATCHING=1 -Oz -g -s ASSERTIONS=2 -s WASM=1 -s "EXPORTED_RUNTIME_METHODS=['removeRunDependency','UTF8ToString','allocate', 'ALLOC_NORMAL', 'ALLOC_STACK', 'ALLOC_DYNAMIC', 'ALLOC_NONE', '_activate_trigger', '_set_toggle_button',  'cwrap', 'ccall', 'FT_Get_Char_Index', 'FT_Load_Gryph', 'FT_New_Memory_Face', 'FT_Render_Glyph', 'FT_Set_Char_Size', 'emscripten_fetch','emscripten_fetch_attr_init', 'emscripten_fetch_close', '_set_float', '_get_integer', '_set_integer', '_set_float', '_set_string', '_set_resize_event', '_stop_music_playing', '_set_background_mode', '_set_new_script']" -s "EXPORTED_FUNCTIONS=['_malloc', '_free', '_main', '_set_float', '_get_integer', '_set_integer', '_set_float', '_set_string', '_set_resize_event', '_stop_music_playing', '_set_background_mode', '_set_new_script']" -s ALLOW_BLOCKING_ON_MAIN_THREAD=1 -s USE_SDL_MIXER=2  -s FILESYSTEM=0   --use-preload-cache  -s NO_EXIT_RUNTIME=1 -s TOTAL_MEMORY=1750Mb  --bind -std=c++11 web_page/test.cpp libGameApi.bc -s USE_SDL=2  -s USE_FREETYPE=1 -L. -Lobs -I. -o web_page/web_page_highmem.js --memory-init-file 0 -s ERROR_ON_UNDEFINED_SYMBOLS=0   -s ENVIRONMENT=web,worker -s DISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR=1 -s FETCH=1  -s MIN_WEBGL_VERSION=2 -s MAX_WEBGL_VERSION=2 -s GL_SUPPORT_AUTOMATIC_ENABLE_EXTENSIONS=0 )

The web_page_highmem.js file does not have asmConsts map at all.

**Comment by sbc100** [comment]
Sorry its called `ASM_CONSTS`

**Comment by terop2** [comment]
var ASM_CONSTS = {
  2379911: ($0) => { var str = UTF8ToString($0) + '\n\n' + 'Abort/Retry/Ignore/AlwaysIgnore? [ariA] :'; var reply = window.prompt(str, "i"); if (reply === null) { reply = "i"; } return allocate(intArrayFromString(reply), 'i8', ALLOC_NORMAL); },  
 2380136: () => { if (typeof(AudioContext) !== 'undefined') { return true; } else if (typeof(webkitAudioContext) !== 'undefined') { return true; } return false; },  
 2380283: () => { if ((typeof(navigator.mediaDevices) !== 'undefined') && (typeof(navigator.mediaDevices.getUserMedia) !== 'undefined')) { return true; } else if (typeof(navigator.webkitGetUserMedia) !== 'undefined') { return true; } return false; },  
 2380517: ($0) => { if(typeof(Module['SDL2']) === 'undefined') { Module['SDL2'] = {}; } var SDL2 = Module['SDL2']; if (!$0) { SDL2.audio = {}; } else { SDL2.capture = {}; } if (!SDL2.audioContext) { if (typeof(AudioContext) !== 'undefined') { SDL2.audioContext = new AudioContext(); } else if (typeof(webkitAudioContext) !== 'undefined') { SDL2.audioContext = new webkitAudioContext(); } if (SDL2.audioContext) { autoResumeAudioContext(SDL2.audioContext); } } return SDL2.audioContext === undefined ? -1 : 0; },  
 2381010: () => { var SDL2 = Module['SDL2']; return SDL2.audioContext.sampleRate; },  
 2381078: ($0, $1, $2, $3) => { var SDL2 = Module['SDL2']; var have_microphone = function(stream) { if (SDL2.capture.silenceTimer !== undefined) { clearTimeout(SDL2.capture.silenceTimer); SDL2.capture.silenceTimer = undefined; } SDL2.capture.mediaStreamNode = SDL2.audioContext.createMediaStreamSource(stream); SDL2.capture.scriptProcessorNode = SDL2.audioContext.createScriptProcessor($1, $0, 1); SDL2.capture.scriptProcessorNode.onaudioprocess = function(audioProcessingEvent) { if ((SDL2 === undefined) || (SDL2.capture === undefined)) { return; } audioProcessingEvent.outputBuffer.getChannelData(0).fill(0.0); SDL2.capture.currentCaptureBuffer = audioProcessingEvent.inputBuffer; dynCall('vi', $2, [$3]); }; SDL2.capture.mediaStreamNode.connect(SDL2.capture.scriptProcessorNode); SDL2.capture.scriptProcessorNode.connect(SDL2.audioContext.destination); SDL2.capture.stream = stream; }; var no_microphone = function(error) { }; SDL2.capture.silenceBuffer = SDL2.audioContext.createBuffer($0, $1, SDL2.audioContext.sampleRate); SDL2.capture.silenceBuffer.getChannelData(0).fill(0.0); var silence_callback = function() { SDL2.capture.currentCaptureBuffer = SDL2.capture.silenceBuffer; dynCall('vi', $2, [$3]); }; SDL2.capture.silenceTimer = setTimeout(silence_callback, ($1 / SDL2.audioContext.sampleRate) * 1000); if ((navigator.mediaDevices !== undefined) && (navigator.mediaDevices.getUserMedia !== undefined)) { navigator.mediaDevices.getUserMedia({ audio: true, video: false }).then(have_microphone).catch(no_microphone); } else if (navigator.webkitGetUserMedia !== undefined) { navigator.webkitGetUserMedia({ audio: true, video: false }, have_microphone, no_microphone); } },  
 2382730: ($0, $1, $2, $3) => { var SDL2 = Module['SDL2']; SDL2.audio.scriptProcessorNode = SDL2.audioContext['createScriptProcessor']($1, 0, $0); SDL2.audio.scriptProcessorNode['onaudioprocess'] = function (e) { if ((SDL2 === undefined) || (SDL2.audio === undefined)) { return; } SDL2.audio.currentOutputBuffer = e['outputBuffer']; dynCall('vi', $2, [$3]); }; SDL2.audio.scriptProcessorNode['connect'](SDL2.audioContext['destination']); },  
 2383140: ($0, $1) => { var SDL2 = Module['SDL2']; var numChannels = SDL2.capture.currentCaptureBuffer.numberOfChannels; for (var c = 0; c < numChannels; ++c) { var channelData = SDL2.capture.currentCaptureBuffer.getChannelData(c); if (channelData.length != $1) { throw 'Web Audio capture buffer length mismatch! Destination size: ' + channelData.length + ' samples vs expected ' + $1 + ' samples!'; } if (numChannels == 1) { for (var j = 0; j < $1; ++j) { setValue($0 + (j * 4), channelData[j], 'float'); } } else { for (var j = 0; j < $1; ++j) { setValue($0 + (((j * numChannels) + c) * 4), channelData[j], 'float'); } } } },  
 2383745: ($0, $1) => { var SDL2 = Module['SDL2']; var numChannels = SDL2.audio.currentOutputBuffer['numberOfChannels']; for (var c = 0; c < numChannels; ++c) { var channelData = SDL2.audio.currentOutputBuffer['getChannelData'](c); if (channelData.length != $1) { throw 'Web Audio output buffer length mismatch! Destination size: ' + channelData.length + ' samples vs expected ' + $1 + ' samples!'; } for (var j = 0; j < $1; ++j) { channelData[j] = HEAPF32[$0 + ((j*numChannels + c) << 2) >> 2]; } } },  
 2384225: ($0) => { var SDL2 = Module['SDL2']; if ($0) { if (SDL2.capture.silenceTimer !== undefined) { clearTimeout(SDL2.capture.silenceTimer); } if (SDL2.capture.stream !== undefined) { var tracks = SDL2.capture.stream.getAudioTracks(); for (var i = 0; i < tracks.length; i++) { SDL2.capture.stream.removeTrack(tracks[i]); } SDL2.capture.stream = undefined; } if (SDL2.capture.scriptProcessorNode !== undefined) { SDL2.capture.scriptProcessorNode.onaudioprocess = function(audioProcessingEvent) {}; SDL2.capture.scriptProcessorNode.disconnect(); SDL2.capture.scriptProcessorNode = undefined; } if (SDL2.capture.mediaStreamNode !== undefined) { SDL2.capture.mediaStreamNode.disconnect(); SDL2.capture.mediaStreamNode = undefined; } if (SDL2.capture.silenceBuffer !== undefined) { SDL2.capture.silenceBuffer = undefined } SDL2.capture = undefined; } else { if (SDL2.audio.scriptProcessorNode != undefined) { SDL2.audio.scriptProcessorNode.disconnect(); SDL2.audio.scriptProcessorNode = undefined; } SDL2.audio = undefined; } if ((SDL2.audioContext !== undefined) && (SDL2.audio === undefined) && (SDL2.capture === undefined)) { SDL2.audioContext.close(); SDL2.audioContext = undefined; } },  
 2385397: ($0, $1, $2) => { var w = $0; var h = $1; var pixels = $2; if (!Module['SDL2']) Module['SDL2'] = {}; var SDL2 = Module['SDL2']; if (SDL2.ctxCanvas !== Module['canvas']) { SDL2.ctx = Module['createContext'](Module['canvas'], false, true); SDL2.ctxCanvas = Module['canvas']; } if (SDL2.w !== w || SDL2.h !== h || SDL2.imageCtx !== SDL2.ctx) { SDL2.image = SDL2.ctx.createImageData(w, h); SDL2.w = w; SDL2.h = h; SDL2.imageCtx = SDL2.ctx; } var data = SDL2.image.data; var src = pixels >> 2; var dst = 0; var num; if (typeof CanvasPixelArray !== 'undefined' && data instanceof CanvasPixelArray) { num = data.length; while (dst < num) { var val = HEAP32[src]; data[dst ] = val & 0xff; data[dst+1] = (val >> 8) & 0xff; data[dst+2] = (val >> 16) & 0xff; data[dst+3] = 0xff; src++; dst += 4; } } else { if (SDL2.data32Data !== data) { SDL2.data32 = new Int32Array(data.buffer); SDL2.data8 = new Uint8Array(data.buffer); SDL2.data32Data = data; } var data32 = SDL2.data32; num = data32.length; data32.set(HEAP32.subarray(src, src + num)); var data8 = SDL2.data8; var i = 3; var j = i + 4*num; if (num % 8 == 0) { while (i < j) { data8[i] = 0xff; i = i + 4 | 0; data8[i] = 0xff; i = i + 4 | 0; data8[i] = 0xff; i = i + 4 | 0; data8[i] = 0xff; i = i + 4 | 0; data8[i] = 0xff; i = i + 4 | 0; data8[i] = 0xff; i = i + 4 | 0; data8[i] = 0xff; i = i + 4 | 0; data8[i] = 0xff; i = i + 4 | 0; } } else { while (i < j) { data8[i] = 0xff; i = i + 4 | 0; } } } SDL2.ctx.putImageData(SDL2.image, 0, 0); },  
 2386866: ($0, $1, $2, $3, $4) => { var w = $0; var h = $1; var hot_x = $2; var hot_y = $3; var pixels = $4; var canvas = document.createElement("canvas"); canvas.width = w; canvas.height = h; var ctx = canvas.getContext("2d"); var image = ctx.createImageData(w, h); var data = image.data; var src = pixels >> 2; var dst = 0; var num; if (typeof CanvasPixelArray !== 'undefined' && data instanceof CanvasPixelArray) { num = data.length; while (dst < num) { var val = HEAP32[src]; data[dst ] = val & 0xff; data[dst+1] = (val >> 8) & 0xff; data[dst+2] = (val >> 16) & 0xff; data[dst+3] = (val >> 24) & 0xff; src++; dst += 4; } } else { var data32 = new Int32Array(data.buffer); num = data32.length; data32.set(HEAP32.subarray(src, src + num)); } ctx.putImageData(image, 0, 0); var url = hot_x === 0 && hot_y === 0 ? "url(" + canvas.toDataURL() + "), auto" : "url(" + canvas.toDataURL() + ") " + hot_x + " " + hot_y + ", auto"; var urlBuf = _malloc(url.length + 1); stringToUTF8(url, urlBuf, url.length + 1); return urlBuf; },  
 2387855: ($0) => { if (Module['canvas']) { Module['canvas'].style['cursor'] = UTF8ToString($0); } },  
 2387938: () => { if (Module['canvas']) { Module['canvas'].style['cursor'] = 'none'; } },  
 2388007: () => { return window.innerWidth; },  
 2388037: () => { return window.innerHeight; }
};

**Comment by sbc100** [comment]
2387855 is contained in your map and looks correct:

```
2387855: ($0) => { if (Module['canvas']) { Module['canvas'].style['cursor'] = UTF8ToString($0); } },
```

Can you attach a debugger and see why `runEmAsmFunction` is not finding it ?

**Comment by terop2** [comment]
its difficult to reproduce the problem since it only happens rarely. Normally it runs ok, but once in every 50 startups it gives this error, probably some kind of timing problem

**Comment by sbc100** [comment]
Thats very odd since the `ASM_CONSTS` is never modified.  I'm not sure why that error would ever be shown.

Is this code running on the main thread?  (i.e. is `GameApi::MainLoopApi::init_window` called on the main thread)?

**Comment by terop2** [comment]
yes, its in main thread.

**Comment by kripken** [comment]
This could be a browser bug. On which browsers do you see it?

**Comment by terop2** [comment]
this is chromium on ubuntu

**Comment by sbc100** [comment]
`-sPTHREAD_POOL_SIZE=30` is quite a lot of workers.. do you need that many?  It could be putting some kind of strain on the browser/system?

**Comment by terop2** [comment]
yes, I think 30 is good number in there, since I'm using threads alot.

**Comment by sbc100** [comment]
This seems like it should be an thing to debug if you can ever catch it in debugger.  Just look at `ASM_CONSTS` and see if/why 2387855 is not there.

**Comment by terop2** [comment]
I debugged it, but cannot figure out the problem. I can see these code pieces:
 function runMainThreadEmAsm(code, sigPtr, argbuf, sync) {
      var args = readEmAsmArgs(sigPtr, argbuf);
      if (ENVIRONMENT_IS_PTHREAD) {
        // EM_ASM functions are variadic, receiving the actual arguments as a buffer
        // in memory. the last parameter (argBuf) points to that data. We need to
        // always un-variadify that, *before proxying*, as in the async case this
        // is a stack allocation that LLVM made, which may go away before the main
        // thread gets the message. For that reason we handle proxying *after* the
        // call to readEmAsmArgs, and therefore we do that manually here instead
        // of using __proxy. (And dor simplicity, do the same in the sync
        // case as well, even though it's not strictly necessary, to keep the two
        // code paths as similar as possible on both sides.)
        // -1 - code is the encoding of a proxied EM_ASM, as a negative number
        // (positive numbers are non-EM_ASM calls).
        return _emscripten_proxy_to_main_thread_js.apply(null, [-1 - code, sync].concat(args));
      }
      if (!ASM_CONSTS.hasOwnProperty(code)) abort('No EM_ASM constant found at address ' + code);
      return ASM_CONSTS[code].apply(null, args);
    }
function _emscripten_asm_const_int_sync_on_main_thread(code, sigPtr, argbuf) {
      return runMainThreadEmAsm(code, sigPtr, argbuf, 1);
    }
 )
  (func $Emscripten_ShowCursor (;34644;) (param $var0 i32) (result i32)
    (local $var1 i32)
    (local $var2 i32)
    (local $var3 i32)
    global.get $__stack_pointer
    i32.const 16
    i32.sub
    local.tee $var1
    local.tee $var2
    global.get $global12
    i32.gt_u
    local.get $var2
    global.get $global13
    i32.lt_u
    i32.or
    if
      local.get $var2
      call $__handle_stack_overflow
    end
    local.get $var2
    global.set $__stack_pointer
    block $label0
      call $SDL_GetMouseFocus
      i32.eqz
      br_if $label0
      block $label1
        local.get $var0
        i32.eqz
        br_if $label1
        local.get $var0
        i32.load offset=4
        local.tee $var0
        i32.eqz
        br_if $label1
        local.get $var0
        i32.load
        local.tee $var0
        i32.eqz
        br_if $label0
        local.get $var1
        local.get $var0
        i32.store
        local.get $var1
        i32.const 105
        i32.store16 offset=14 align=1
        i32.const 2387855
        local.get $var1
        i32.const 14
        i32.add
        local.get $var1
        call $emscripten_asm_const_int_sync_on_main_thread
        drop
        br $label0
      end $label1
      local.get $var1
      i32.const 0
      i32.store8 offset=13
      i32.const 2387938
      local.get $var1
      i32.const 13
      i32.add
      i32.const 0
      call $emscripten_asm_const_int_sync_on_main_thread
      drop
    end $label0
    local.get $var1
    i32.const 16
    i32.add
    local.tee $var3
    global.get $global12
    i32.gt_u
    local.get $var3
    global.get $global13
    i32.lt_u
    i32.or
    if
      local.get $var3
      call $__handle_stack_overflow
    end
    local.get $var3
    global.set $__stack_pointer
    i32.const 0
  )

But no idea why ASM_CONSTS is not having that line? Is it coming too late in the initialization, since this is very early in the main() function?

**Comment by terop2** [comment]
But maybe it needs some other way to handle the error situation than just abort()? If it just ignored it, it would solve the issue?

**Comment by terop2** [comment]
I'm not sure if I would care if the cursor does some funny stuff, but emscripten app crashing is kinda bad...

**Comment by sbc100** [comment]
No, this is not a recoverable error.  If `ASM_CONSTS` doesn't have that value some really bad/fatal has gone wrong.

Can you see the `ASM_CONSTS` global in the debugger?  Can you `console.log(ASM_CONSTS[2387855])` ?  If not can you run `console.log(ASM_CONSTS)` to see what `ASM_CONSTS` does contain?

**Comment by terop2** [comment]
![logpic](https://github.com/emscripten-core/emscripten/assets/18078535/84caafbd-fd02-45f6-a29a-f0af6f34d0fb)

I didn't reproduce the problem yet, but ASM_CONSTS definitely have something wrong before main is executed, as all items are not yet filled for it. the 2387855 is not there.

**Comment by terop2** [comment]
oh wait, its in the picture.

**Comment by terop2** [comment]
I think it updated the picture later

**Comment by terop2** [comment]
big problem is that reproducing the actual error is difficult/it takes a while to reproduce.

**Comment by sbc100** [comment]
I assume that screenshot was not from after the crash/abort?  (since 2387855 exists in ASM_CONSTS)

**Comment by terop2** [comment]
yes I havent got the error reproduced today any longer. But it randomly appears sometimes.

**Comment by terop2** [comment]
Got it reproduced:
![l1](https://github.com/emscripten-core/emscripten/assets/18078535/ad8d2cdc-d75a-4fbb-a883-70c07714a27c)
![l2](https://github.com/emscripten-core/emscripten/assets/18078535/23bfdbc9-fd45-4535-be1a-1cc4efe74268)

That shows the situation, i.e. the number isn't in the ASM_CONSTS.

**Comment by terop2** [comment]
But I think I know why it is difficult to reproduce. Happens only after I have recompiled the whole emscripten/pushed the wasm files to the web server => could be wasm file caching problem?

**Comment by terop2** [comment]
also looking at the id numbers, it looks like the number requested are not even in the range of values where the implementations are at? Maybe it generates those id numbers twice, and uses the old values?

**Comment by sbc100** [comment]
Yes, this is almost certainly an issue where the JS file and the wasm file are not in sync.  i.e. the browser is setting the new JS, or the new wasm but not both.

**Comment by sbc100** [comment]
I would check your server settings, in particular the cache control ones: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control

**Comment by terop2** [comment]
I changed the server config a little, now it's "Cache-Control: max-age=604800, immutable", and we'll see if the errors are being fixed by that change.

**Comment by terop2** [comment]
NOTE: the immutable requires that the .wasm and .js file names are being changed by the modification date of the files, like foo.wasm?378634628 kind of system.

**Comment by juj** [comment]
The `No EM_ASM constant found at address x` error is most commonly caused by a stale cache. The .js file is small enough that the browser caches it, but on a larger application the browser will refuse to cache the .wasm file, which will result in the browser doing different caching strategies between the two files, resulting in mismatched files being fetched.

Try holding down the Shift key when clicking on page Reload button or the F5 key to force the browser to fetch a new copy from the web server. That can help diagnose issues that are caused by caching.

**Comment by terop2** [comment]
hmm, my immutable cache fix didn't help with the problem. I'm still getting EM_ASM crashes :(

**Comment by sbc100** [comment]
This is clearly a resource mismatch issue though right?   Using ctrl-shift reload always fixed right?

**Comment by terop2** [comment]
yes, reloading the page always fixes it

**Comment by terop2** [comment]
I figured out a fix that works: removing -g from compile flags makes wasm size small enough that it fixes the issue.

**Comment by terop2** [comment]
i got better solution by using -gsplit-dwarf, it keeps wasm file size at 5Mb, but rest is in dwp file...

**Comment by terop2** [comment]
I think this error can be closed now. I have everything I need for this.
